### ISSUE TITLE
Bugs/vulnerabilities found in hicolor v0.5.0

### ISSUE THREAD

**Issue by Helson-S** [dbohdan/hicolor#5]
# Summary

Hi~，I did some fuzzy testing and found some bugs/vulnerabilities on hicolor v0.5.0. I hope these findings will help improve software quality.

These bugs/vulnerabilities are mainly caused by unsafe component cute_png.h v1.05. According to my analysis, Because the compilation environment of hicolor is inconsistent with the official compilation environment of cute_png.h v1.05, not all bugs in cute_png.h affect hicolor. The bugs/vulnerabilities listed below can truly affect hicolor v0.5.0.

All of the bugs/vulnerablities are triggered with no assertion raised. This means that these bugs/vulnerabilities are unexpected behaviors of the program. 

hicolor:  https://github.com/dbohdan/hicolor

cute_headers: https://github.com/RandyGaul/cute_headers

See also https://github.com/Helson-S/FuzzyTesting/tree/master/hicolor

# heapof-r1-cp_unfilter-cute_png-1019c11
## Description

Heap-buffer-flow bug/vulnerability caused by read access found in function cp_unfilter() at line 1019 of vendor/cute_png.h v1.05.

Affected version: hicolor v0.5.0

## Reproduction

Environment:

Operating system version: Ubuntu 22.04

Linux kernel version: Linux pc 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Compiler version: gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

Compile hicolor with ASAN and run the following command in bash shell:

```shell
hicolor encode -a ./poc/sample6.png ./output.hic && rm -f ./output.hic
```

## Screen-shot
![image-20240530183857985](https://github.com/dbohdan/hicolor/assets/49753907/3ca2fda9-2f21-414e-8d86-a1759a71ae61)

# heapof-r65280-cp_stored-cute_png-543c2
## Description

Heap-buffer-flow bug/vulnerability caused by read access found in function cp_stored() at line 543 of vendor/cute_png.h v1.05. What's more, sample10.png provided as attack vector causes double-free heap memory corruption in function cp_load_png_mem() at line 1194 of vendor/cute_png.h v1.05.

Affected version: hicolor v0.5.0

## Reproduction

Environment:

Operating system version: Ubuntu 22.04

Linux kernel version: Linux pc 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Compiler version: gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

Compile hicolor with ASAN and run the following command in bash shell:

```shell
hicolor encode -a ./poc/sample10.png ./output.hic && rm -f ./output.hic
```

### heap-buffer-overflow
![image-20240530184723547](https://github.com/dbohdan/hicolor/assets/49753907/321670ef-1051-4d98-9952-f59eb6b6c4b6)

### double-free heap memory corruption
![image-20240530184848743](https://github.com/dbohdan/hicolor/assets/49753907/328bc6b1-4511-4682-84ca-c5cd99be72d0)

![image-20240530185015780](https://github.com/dbohdan/hicolor/assets/49753907/bbaea8f4-5662-481a-8100-a032733c1e48)

# heapof-w1-cp_block-cute_png-623c12

## Description

Heap-buffer-flow bug/vulnerability caused by write access found in function cp_block() at line 623 of vendor/cute_png.h v1.05. What's more, sample11.png provided as attack vector causes double-free heap memory corruption in function cp_load_png() at line 1216 of vendor/cute_png.h v1.05.

Affected version: hicolor v0.5.0

## Reproduction

Environment:

Operating system version: Ubuntu 22.04

Linux kernel version: Linux pc 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Compiler version: gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

Compile hicolor with ASAN and run the following command in bash shell:

```shell
hicolor encode -a ./poc/sample11.png ./output.hic && rm -f ./output.hic
```

### heap-buffer-overflow
![image-20240530185401405](https://github.com/dbohdan/hicolor/assets/49753907/db00252b-d612-4208-97dc-6bd6114c13bf)

### double-free heap memory corruption
![image-20240530185451914](https://github.com/dbohdan/hicolor/assets/49753907/c371740f-d9a9-4293-8f43-6c1b709c7062)

![image-20240530185539214](https://github.com/dbohdan/hicolor/assets/49753907/94de57ba-6fb2-40ad-9083-fe168aa7d8d7)

# heapof-w1-png_quantize-cli-220c32
## Description

heap-buffer-overflow bug/vulnerability caused by write access found in function png_quantize() at line 220 of cli.c v1.05.

Affected version: hicolor v0.5.0

## Reproduction

Environment:

Operating system version: Ubuntu 22.04

Linux kernel version: Linux pc 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Compiler version: gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

Compile hicolor with ASAN and run the following command in bash shell:

```shell
hicolor quantize -n ./poc/sample18.png ./output.hic && rm -f ./output.hic
```

## Screen-shot
![image-20240530225208577](https://github.com/dbohdan/hicolor/assets/49753907/b2dd55cc-71ab-45c5-952c-74a6c4bb5395)

# heapof-w16-cp_block-cute_png-644c37
## Description

Heap-buffer-flow bug/vulnerability caused by write access found in function cp_block() at line 644 of vendor/cute_png.h v1.05. What's more, sample12.png provided as attack vector causes unmap invalid pointer memory corruption in function cp_load_png_mem() at line 1189 of vendor/cute_png.h v1.05.

Affected version: hicolor v0.5.0

## Reproduction

Environment:

Operating system version: Ubuntu 22.04

Linux kernel version: Linux pc 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Compiler version: gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

Compile hicolor with ASAN and run the following command in bash shell:

```shell
hicolor encode -a ./poc/sample12.png ./output.hic && rm -f ./output.hic
```

### heap-buffer-overflow
![image-20240530192030403](https://github.com/dbohdan/hicolor/assets/49753907/3261c734-54ba-4f98-830f-762fb44f016f)

### unmap invalid pointer
![image-20240530192010075](https://github.com/dbohdan/hicolor/assets/49753907/f26e51ad-7d0b-43ee-b114-942e9ec0d24b)

![image-20240530192345688](https://github.com/dbohdan/hicolor/assets/49753907/eec275bb-df97-4d7d-b4ed-6fd7a7e760d2)

# heapof-w98-cp_block-5c0-cute_png-642c5
## Description

Heap-buffer-flow bug/vulnerability caused by write access found in function cp_block() at line 642 of vendor/cute_png.h v1.05.

Affected version: hicolor v0.5.0

## Reproduction

Environment:

Operating system version: Ubuntu 22.04

Linux kernel version: Linux pc 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Compiler version: gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

Compile hicolor with ASAN and run the following command in bash shell:

```shell
hicolor encode -a ./poc/sample13.png ./output.hic && rm -f ./output.hic
```

## Screen-shot
![image-20240530192505615](https://github.com/dbohdan/hicolor/assets/49753907/dde37c3d-91f4-4a3b-b707-76ebff776bfc)

![image-20240531002753478](https://github.com/dbohdan/hicolor/assets/49753907/a7b36f2d-5c0c-496f-917f-72dead4796db)

# stkof-w133-cp_dynamic-cute_png-603
## Description

stack-buffer-overflow bug/vulnerability caused by write access found in function cp_dynamic() at line 603 of vendor/cute_png.h v1.05. It will lead to control flow hijacking.

Affected version: hicolor v0.5.0

## Reproduction

Environment:

Operating system version: Ubuntu 22.04

Linux kernel version: Linux pc 5.19.0-41-generic #42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

Compiler version: gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04)

Compile hicolor with ASAN and run the following command in bash shell:

```shell
hicolor encode -a ./poc/sample16.png ./output.hic && rm -f ./output.hic
```

## Screen-shot
![image-20240530223831738](https://github.com/dbohdan/hicolor/assets/49753907/6989ee39-704c-46d2-a699-77c411479b65)

![image-20240530223921086](https://github.com/dbohdan/hicolor/assets/49753907/3c6d108e-387d-4afe-83f1-59848b4a68c3)

**Comment by dbohdan** [comment]
Thanks for the report. I will look into the stack overflow in `png_quantize`. As for cute_png, like https://github.com/RandyGaul/cute_headers/issues/381#issuecomment-2140439440 says, it is not designed for untrusted input. I should note this in the readme. I may eventually address the insecurity of cute_png by replacing it with another library.

An accessibility suggestion: it would be better if your screenshots were code blocks. If you don't want code blocks making your issue too long, hide them inside [`<details>` tags](https://gist.github.com/ericclemmons/b146fe5da72ca1f706b2ef72a20ac39d).

**Comment by dbohdan** [comment]
Fixed in bf9c0623406509cdd602e4c1db70f858db84024f by switching from cute_png to libpng.
